### PR TITLE
Prepare vmm/builder for building from snapshot

### PR DESCRIPTION
**Pull request by acatangiu**
## Reason for This PR

Prerequisite for #1847 

## Description of Changes

Rearranges vmm creation code to differentiate between creating resources and configuring them for boot.

This is a prerequisite for building from snapshot in order to share logic between the two paths.

**Note**: should be much easier to review going one commit at a time.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: src/vmm/src/builder.rs [hunk 1]
```diff
@@ -622,12 +589,20 @@ fn create_vcpus_aarch64(
 /// Configures the system for booting Linux.
 pub fn configure_system_for_boot(
     vmm: &Vmm,
-    vcpus: &[Vcpu],
+    vcpus: &mut [Vcpu],
+    _vcpu_config: VcpuConfig,
```
Comment (sandreim) [inline src/vmm/src/builder.rs:576]: Is there a better way to make this parameter x86_64 specific and not work around compiler unused warning ?

File: src/vmm/src/builder.rs [hunk 2]
```diff
@@ -508,30 +506,15 @@ pub fn setup_serial_device(
...
-                })
-                .map_err(StartMicrovmError::Internal)?;
-        }};
-    }
-
-    register_irqfd_evt!(com_evt_1_3, 4);
-    register_irqfd_evt!(com_evt_2_4, 3);
-    register_irqfd_evt!(kbd_evt, 1);
-    Ok(())
+    serial: Arc<Mutex<devices::legacy::Serial>>,
```
Comment (sandreim) [inline src/vmm/src/builder.rs:509]: The naming is not consistent with impl: creates a PIO device manager and also attaches devices.

File: src/vmm/src/device_manager/legacy.rs [hunk 3]
```diff
@@ -113,23 +114,39 @@ impl PortIODeviceManager {
...
+        macro_rules! register_irqfd_evt {
+            ($evt: ident, $index: expr) => {{
+                vm_fd
+                    .register_irqfd(&self.$evt, $index)
+                    .map_err(|e| Error::EventFd(std::io::Error::from_raw_os_error(e.errno())))?;
+            }};
+        }
+        register_irqfd_evt!(com_evt_1_3, 4);
+        register_irqfd_evt!(com_evt_2_4, 3);
+        register_irqfd_evt!(kbd_evt, 1);
```
Comment (sandreim) [inline src/vmm/src/device_manager/legacy.rs:127]: There are 10 loc here, if we get rid of the macro there will be 9. Can we remove the macro? WDYT ?

File: src/vmm/src/builder.rs [hunk 4]
```diff
@@ -1205,7 +1223,13 @@ pub mod tests {
         let tmp_sock_file = TempSockFile::new(TempFile::new().unwrap());
         let vsock_config = default_config(&tmp_sock_file);
 
-        insert_vsock_device(&mut vmm, &mut event_manager, vsock_config);
+        let mut cmdline = default_kernel_cmdline();
+        insert_vsock_device(&mut vmm, &mut cmdline, &mut event_manager, vsock_config);
+        // Check if these the vsock device is described in kernel_cmdline.
```
Comment (ioanachirca) [inline src/vmm/src/builder.rs:1148]: nit:
```suggestion
        // Check if the vsock device is described in kernel_cmdline.
```

File: src/vmm/src/device_manager/mmio.rs [hunk 5]
```diff
@@ -92,6 +92,9 @@ impl MMIODeviceManager {
     /// Create a new DeviceManager handling mmio devices (virtio net, block).
     pub fn new(mmio_base: &mut u64, irq_interval: (u32, u32)) -> MMIODeviceManager {
         if cfg!(target_arch = "aarch64") {
+            // This hack is done to reserve the 'log boot time magic address'.
+            // TODO (issue #1865): we should create a simple mmio device for logging the boot time
```
Comment (lauralt) [inline src/vmm/src/device_manager/mmio.rs:96]: :+1:

File: src/vmm/src/builder.rs (hunk 2 above)
Comment (acatangiu) [reply src/vmm/src/builder.rs:509]: done

File: src/vmm/src/device_manager/mmio.rs [hunk 6]
```diff
@@ -291,7 +291,7 @@ mod tests {
     const QUEUE_SIZES: &[u16] = &[64];
 
     impl MMIODeviceManager {
-        fn register_virtio_device(
+        fn register_virtio_helper(
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply src/vmm/src/device_manager/mmio.rs:267]: done, but didn't go back to the original, I want a name that is explicit about the fact that it is a `tests` function

File: src/vmm/src/device_manager/mmio.rs [hunk 7]
```diff
@@ -70,6 +70,15 @@ const MMIO_LEN: u64 = 0x1000;
 /// to its configuration space.
 pub const MMIO_CFG_SPACE_OFF: u64 = 0x100;
 
+/// Private structure for storing information about the MMIO device registered at some address
+/// on the bus.
+#[derive(Clone, Debug)]
+pub struct MMIODeviceInfo {
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply src/vmm/src/device_manager/mmio.rs:76]: done

File: src/vmm/src/device_manager/legacy.rs (hunk 3 above)
Comment (acatangiu) [reply src/vmm/src/device_manager/legacy.rs:127]: done

**Comment by acatangiu** [comment]
@sandreim @ioanachirca @lauralt I've addressed all your comments, please take another look.

**Pull request merged**